src/goto-instrument/contracts/instrument_spec_assigns.h

```diff
@@ -0,0 +1,354 @@
+  {
+    return operands()[0];
+  }
+
+  /// Target expression
+  const exprt &target() const
+  {
+    return operands()[1];
+  }
+};
+
+/// Class that represents a normalized conditional address range, with:
+/// - condition expression
+/// - target expression
+/// - normalised start address expression
+/// - normalised size expression
+/// - snapshot variable symbols
+class car_exprt : public exprt
+{
+public:
+  car_exprt(
+    const exprt &_condition,
+    const exprt &_target,
+    const exprt &_target_start_address,
+    const exprt &_target_size,
+    const symbol_exprt &_validity_var,
+    const symbol_exprt &_lower_bound_var,
+    const symbol_exprt &_upper_bound_var,
+    const source_locationt &_source_location)
+    : exprt()
+  {
+    add_to_operands(_condition);
+    add_to_operands(_target);
+    add_to_operands(_target_start_address);
+    add_to_operands(_target_size);
+    add_to_operands(_validity_var);
+    add_to_operands(_lower_bound_var);
+    add_to_operands(_upper_bound_var);
+    add_source_location() = _source_location;
+  }
+
+  /// Condition expression. When this condition holds the target is allowed
+  const exprt &condition() const
+  {
+    return operands()[0];
+  }
+
+  /// The target expression
+  const exprt &target() const
+  {
+    return operands()[1];
+  }
+
+  /// Start address of the target
+  const exprt &target_start_address() const
+  {
+    return operands()[2];
+  }
+
+  /// Size of the target in bytes
+  const exprt &target_size() const
+  {
+    return operands()[3];
+  }
+
+  /// Identifier of the validity snapshot variable
+  const symbol_exprt &valid_var() const
+  {
+    return to_symbol_expr(operands()[4]);
+  }
+
+  /// Identifier of the lower address bound snapshot variable
+  const symbol_exprt &lower_bound_var() const
+  {
+    return to_symbol_expr(operands()[5]);
+  }
+
+  /// Identifier of the upper address bound snapshot variable
+  const symbol_exprt &upper_bound_var() const
+  {
+    return to_symbol_expr(operands()[6]);
+  }
+};
+
+/// \brief A class that generates instrumentation for assigns clause checking.
+///
+/// The `track_*` methods add targets to the sets of tracked targets and append
+/// instructions to the given destination program.
+///
+/// The `check_inclusion_*` methods generate assertions checking for inclusion
+/// of a designated target in one of the tracked targets,
+/// and append instructions to the given destination.
+class instrument_spec_assignst
+{
+public:
+  /// \brief Class constructor.
+  ///
+  ///  \param _function_id name of the instrumented function
+  ///  \param _functions other functions of the model
+  ///  \param _ns namespace of the goto_model
+  ///  \param _st symbol table of the goto_model
+  ///  \param _message_handler message handler used for warning/error messages
+  instrument_spec_assignst(
+    const irep_idt &_function_id,
+    const goto_functionst &_functions,
+    const namespacet &_ns,
+    symbol_tablet &_st,
```

Comment:
Do you expect `_ns` to be the namespace wrapping `_st`? If so, I'd refrain from taking `_ns` here to avoid this implicit contract. Instead, construct your own `namespacet`.